create first draft of branching and release process (#3)
project-codeflare/adr: PCF-ADR-0002-release-story-and-branch-maintenance.md

frequency, and team collaboration.

## Why

We have a few projects now with no accepted development strategy and release cadence. We need to make sure that
developers have an idea of how to get changes into the main branch and when they can expect changes to be released into
a new version as well as ways of pulling bugfixes into existing releases.

## Goals

- Create a branching and release strategy that is followed across all of our sub projects
- Enable flexible release schedules as some projects may have a different release cadence from others.
- A branching strategy which allows for CI/CD .
- A way to add bugfixes to existing releases.

## How

### Decision

We have decided to use GitHub Flow as our branching strategy. GitHub Flow is a simple and lightweight model that follows
these principles:

- The `main` branch is always deployable and represents the latest version of the code.
- New features and bug fixes are developed in short-lived branches, created from the `main` branch.
- When a branch is ready for review, a pull request is opened to merge it into the `main` branch.

> - After the pull request is approved by at least one reviewer, it is merged into the `main` branch and deployed to

Reviewer: What does "deployed to production" mean in our context?

Author: Deployed to production probably isn't accurate in our case, probably just pushed to the `latest` update channels for the operator and the latest tag for controller images.

### How to pull in bug fixes into a release branch

Sometimes, we may need to create a release branch from the `main` branch to prepare for a major or minor release. For
example, we may want to freeze the feature set for a certain version and only allow bug fixes on that branch. In that
case, we can use the following steps to pull in bug fixes from the `main` branch into the release branch:

- Create a release branch from the `main` branch with a name that indicates the minor version number, such as `v1.2`.
- Continue developing new features and bug fixes on the `main` branch as usual.
- If a bug fix on the `main` branch needs to be included in the release branch, cherry-pick the commit that contains the
  fix onto the release branch.
- When the release branch is ready for deployment, create a tag on GitHub with the patch version incremented the, such
  as `v1.2.1`. This will trigger the deployment pipeline for that tag.

### Release process

The release process for our project is very simple and consists of creating a version tag on GitHub. For example, if we
want to release version 1.3 of our software, we just need to create a tag named `v1.3` on GitHub from the latest commit

> on the `main` branch. This will trigger the build and deployment pipelines, or someone will manually take the required

Reviewer: I'm a little bit confused on the use of release branches and tags. This line states that we would create release tags off of main, but that seems to conflict with the existence of release branches. Would we not want to tag off of release branches? If yes, would we always tag at the point of a release? Can a given release branch have multiple tags (i.e. if we need to do multiple builds/releases for a given version

Author: Release branches would only exist to pull in bug fixes to already released versions. At the time of release you would just create a tag like `v0.1.0`, then if there are some bug fixes that we want to patch onto the minor version we create a branch `release-v0.1` from the `v0.1.0` tag. Then cherry pick commits that we want to include in `v0.1.1`. Once all the commits have been pulled in then we can create a release tag `v0.1.1` off of the `release-v0.1` branch. The idea is that feature and bug dev can happen at the same time without having to pull new features into a patch release.

Author: I'll work on the wording

Reviewer: This makes a lot of sense. Thanks for the clarification!

steps for generating build artifacts, etc..

We do not need to create major version tags unless there is a significant change in the software architecture or
functionality that requires a major version bump.

## Non Goals

- Dictating how to build code for release (prefer automation, but a manual process is also okay)
- Setting a release cycle for the codeflare project.
- Creating a versioning scheme for projects.

## Open questions

> - What artifacts should be built to go alongside the releases? For example, should we generate documentation, binaries,

Reviewer: I think at a minimum we should include a changelog. This can go along with the bullet point below - if there are breaking changes we can explicitly list them in the changelog.

Author: We should also probably at the very least zip up the source code so that the code can be downloaded without using `git`

  packages, etc.?

> - How should we handle breaking changes or backward-incompatible changes in our software? Should we use semantic

Reviewer: I vote for SemVer

Author: I agree. I think the only other sane option is calendar versioning, but that doesn't really jive with the idea of patch releases and doesn't capture potentially breaking changes the way semver does.

frequency, and team collaboration.

## Why

We have a few projects now with no accepted development strategy and release cadence. We need to make sure that
developers have an idea of how to get changes into the main branch and when they can expect changes to be released into
a new version as well as ways of pulling bugfixes into existing releases.

## Goals

- Create a branching and release strategy that is followed across all of our sub projects
- Enable flexible release schedules as some projects may have a different release cadence from others.
- A branching strategy which allows for CI/CD .
- A way to add bugfixes to existing releases.

## How

### Decision

We have decided to use GitHub Flow as our branching strategy. GitHub Flow is a simple and lightweight model that follows
these principles:

- The `main` branch is always deployable and represents the latest version of the code.
- New features and bug fixes are developed in short-lived branches, created from the `main` branch.
- When a branch is ready for review, a pull request is opened to merge it into the `main` branch.

> - After the pull request is approved by at least one reviewer, it is merged into the `main` branch and all necessary

Reviewer 2: I think we could clarify this a little, the prow plugin we are using will require us to have an `approved` label and `lgtm` label on PRs. In general the preference is for those two labels to come from different people.

  high quality and fast feedback.
- It encourages collaboration and code review. Pull requests are the central place for discussing and reviewing code
  changes, ensuring that every change is approved by at least one other developer before merging.
- It allows us to release frequently and incrementally. We can deploy new features and bug fixes as soon as they are
  ready, without waiting for a predefined release cycle or schedule.

### How to pull in bug fixes into a release branch

Sometimes, we may need to create a release branch from the `main` branch to prepare for a major or minor release. For
example, we may want to freeze the feature set for a certain version and only allow bug fixes on that branch. In that
case, we can use the following steps to pull in bug fixes from the `main` branch into the release branch:

- Create a release branch from the `main` branch with a name that indicates the minor version number, such as `v1.2`.
- Continue developing new features and bug fixes on the `main` branch as usual.
- If a bug fix on the `main` branch needs to be included in the release branch, cherry-pick the commit that contains the
  fix onto the release branch.

> - When the release branch is ready for deployment, create a tag on GitHub with the patch version incremented the, such

Reviewer 2: ```suggestion
- When the release branch is ready for deployment, create a tag on GitHub with the patch version incremented. For example, `v1.2.0` will become v1.2.1`. This will trigger the deployment pipeline for that tag.
```

>   as `v1.2.1`. This will trigger the deployment pipeline for that tag.

Reviewer 2: ```suggestion
```

### Release process

The release process for our project is very simple and consists of creating a version tag on GitHub. For example, if we

> want to release version 1.3 of our software, we just need to create a tag named `v1.3` on GitHub from the latest commit

Reviewer 2: Just to further highlight our usage of semver, I'd change references of 1.3 to 1.3.0, etc.

on the `main` branch. This will trigger the build and deployment pipelines, or someone will manually take the required
steps for generating build artifacts, etc..

We do not need to create major version tags unless there is a significant change in the software architecture or
functionality that requires a major version bump.

### Versioning Scheme

All projects should use semantic versioning scheme as it gives the most warning for potentially breaking changes when
compared to other schemes such as calendar versioning.

## Non Goals

- Dictating how to build code for release (prefer automation, but a manual process is also okay)
- Setting a release cycle for the codeflare project.

## Open questions

- What artifacts should be built to go alongside the releases? For example, should we generate documentation, binaries,
  packages, etc.?

> - How should we communicate our releases to our stakeholders and customers? Should we use changelogs, release notes,

Reviewer 2: Is there a CodeFlare mail list? At a minimum, we can share these out to the slack channel

Reviewer 3: there is a projectcodeflare@gmail.com account unused currently, also Carlos has the codeflare.dev domain so a new email could be created in that (if we decide that Slack isn't enough as-is atm)

Reviewer 2: I'll start a thread in slack too.

  newsletters, etc.?

## Alternatives

### Git Flow

GitFlow is no longer a popular branching and release strategy as it requires more overhead with branch management and

> more involved merging. One benefit that it does offers is a slower dev process for junior devs.

Reviewer 2: ```suggestion
more involved merging. One benefit that it does offer is a slower dev process for junior devs.
```

### Trunk Based Development (TBD)

TBD and GitHub flow are very similar in that they treat the main branch as a deployable branch that all feature branches
and bug fixes should start from. The main difference comes with expectations on how often developers pull changes into
the repo. TBD discourages branches living longer than a couple of days, any features that are unfinished or untested are
gated using feature flags. GitHub Flow encourages thorough testing before merging anything into the main branch. TBD is
difficult to implement in an open source community where many developers don't have direct access to the main branch, or
an easy way to get one on one time with a maintainer to do a code review.

## Risks

GitHub Flow and Trunk Based Development both require discipline keeping tests up to date and ensuring that the `main`
branch is always in a deployable state.

## Stakeholder Impacts

| Group                         | Key Contacts     | Date       | Impacted? |
| ----------------------------- | ---------------- | ---------- | --------- |

> | CodeFlare SDK                 | key contact name | 4/19/2023  | yes       |
> | MCAD                          | key contact name | 4/19/2023  | yes       |
> | InstaScale                    | key contact name | 4/19/2023  | yes       |
> | CodeFlare Operator            | key contact name | 4/19/2023  | yes       |

Reviewer 2: ```suggestion
| CodeFlare SDK                 | Mustafa Eyceoz | 4/19/2023  | yes       |
| MCAD                          | Abhishek Malvankar | 4/19/2023  | yes       |
| InstaScale                    | Abhishek Malvankar | 4/19/2023  | yes       |
| CodeFlare Operator            | Anish Asthana | 4/19/2023  | yes       |
```

## References

* https://docs.github.com/en/get-started/quickstart/github-flow
* https://trunkbaseddevelopment.com/
* https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow

## Reviews

| Reviewed by                   | Date       | Notes |
| ----------------------------- | ---------  | ------|

> | name                          | date       | ?     |

Reviewer 2: We should add at least Alex, myself, Mustafa, and Abhishek to this list of reviewers. Additional people are welcome to review.

case, we can use the following steps to pull in bug fixes from the `main` branch into the release branch:

- Create a release branch from the `main` branch with a name that indicates the minor version number, such as `v1.2` or 
  `release-v1.2`.
- Continue developing new features and bug fixes on the `main` branch as usual.
- If a bug fix on the `main` branch needs to be included in the release branch, cherry-pick the commit that contains the
  fix onto the release branch.
- When the branch is ready for a new release, create a tag on GitHub with the patch version incremented. For example if
  `v1.2.0` is the last released version for `v1.2`, then the new tag will be `v1.2.1`. This will trigger the release
  pipeline for that tag.

### Release process

The release process for our project is very simple and consists of creating a version tag on GitHub. For example, if we
want to release version 1.3 of our software, we just need to create a tag named `v1.3.0` on GitHub from the latest
commit on the `main` branch. This will trigger the build and release pipelines, or someone will manually take the
required steps for generating build artifacts, etc..

> 

Reviewer 4: We may want to store the artifacts on a stagging area for a few days for people to download and validate as a last gate before going out.

Author: I think this is definitely doable. One could create `v1.1.0-alpha` tag and then pull any bug fixes that are made into the `v1.1` branch as normal. The second tag on that branch would be `v1.1.0` or even `v1.0.0-beta`. I'll add this to the dev process as an option for staging releases.

Reviewer 4: Looks good. Well explained.

We do not need to create major version tags unless there is a significant change in the software architecture or
functionality that requires a major version bump.

### Versioning Scheme

All projects should use semantic versioning scheme as it gives the most warning for potentially breaking changes when
compared to other schemes such as calendar versioning.

## Non Goals

- Dictating how to build code for release (prefer automation, but a manual process is also okay)

> - Setting a release cycle for the codeflare project.

Reviewer 4: ```suggestion
- Setting a release cycle for the Codeflare project.
```
